---
title: "normr: normR obeys mixtures Rules - Normalization & Difference Calling
in ChIP-seq Data"
author: "Johannes Helmuth"
date: "`r Sys.Date()`"
link: https://www.bioconductor.org/packages/release/bioc/html/normr.html
output:
   BiocStyle::html_document:
      toc: true
vignette: >
  %\VignetteIndexEntry{Normalization of ChIP-seq Data}
  %\VignetteEngine{rmarkdown::render}
  \usepackage[utf8]{inputenc}
---

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```

# Introduction to normR

Chromatin immunoprecipitation followed by sequencing (ChIP-seq) provides
genome-wide localization data for DNA-associated proteins. To infer the
regions bound by such proteins the read densities obtained by such a ChIP-seq
experiment are compared to the corresponding read profile obtained by a
control experiment. A meaningful comparison requires normalization to
mitigate the effects of technical biases, e.g. different sequencing depth.
But more importantly the effect of the enrichment of certain regions on the
overall read statistics. Normalization requires knowledge of the regions that
remained unchanged, such that normalization and difference calling are
inseparable.

This package, normR (**n**ormR **o**beys **r**egime **m**ixture **R**ules),
follows this logic and performs normalization and difference calling
simultaneously to identify genomic regions enriched by the ChIP-procedure
(`normr::enrichR()`). In addition, normR enables the comparison between ChIP-seq
data obtained from different conditions allowing for unraveling genomic regions
that change their association with the ChIP-target (`normr::diffR()`). Lastly,
normR is capable to differentiate multiple regimes of enrichment, *i.e.* broad
domains and sharp peaks (`normr::regimeR`).

This vignette explains a common workflow of normrR analysis on NGS data for
calling enrichment, identification of differential ChIP-seq enrichment and
stratification of enrichment regimes. Herein, we provide examples for the export of
results to common data formats like bigWig and bed. We show how analysis
statistics and diagnostic plots are helpful for studying results. In the last
section, we cover more advanced topics including fitting on defined regions,
*e.g.* promoters, or integrating CNV information in differential ChIP-seq
enrichment calls.


## Calling Enrichment with an Input Control

In this first section, we would like to call regions significantly enriched
for reads in the ChIP-seq experiment given a Control alignment. Here, we analyze
ChIP-seq data for both H3K4me3 (pointy enrichment) and H3K27me3 (broad
enrichment) given an Input-seq control alignment originating from a human
immortalised myelogenous leukemia line (K562). Using normR, we show that our
representative region on human chromsome 1 (`chr1:22500000-25000000`) is
enriched for H3K4me3 mostly at promoters and precludes H3K27me3 enrichment which
is most apparent in inter-genic regions.

### Setting up toy data

As part of the normR package, we provide 3 stripped alignment files
containing reads for chr1:22500000-25000000. One Input-seq alignment serves
as control for two ChIP-seq experiments for H3K27me3 and H3K4me3.

```{r}
#Paths to files containing alignment files
path_Input <- system.file("extdata", "K562_Input.bam", package="normr")
path_ChIP_H3K27me3 <- system.file("extdata", "K562_H3K27me3.bam", package="normr")
path_ChIP_H3K4me3 <- system.file("extdata", "K562_H3K4me3.bam", package="normr")
```

That is easy. Now, we have to define our genome sequences: The

```{r}
#A toy genome definition
genome <- data.frame("seqname"="chr1", "length"=25000000)
```

```{r}
#Calling enrichment with enrichR()
enrichment_H3K27me3 <- enrichR(path_ChIP_H3K27me3, path_Input, genome, procs=1)

#Giving some fitting results
summary(enrichment_H3K27me3)
```

```{r}
#Calling enrichment with enrichR()
enrichment_H3K4me3 <- enrichR(path_ChIP_H3K4me3, path_Input, genome, procs=1)

#Giving some fitting results
summary(enrichment_H3K4me3)
```

```{r, fig.show='hold', fig.cap="Density of enrichment", fig.width=5, fig.height=5}
par(mfrow=c(1,2), bty="n")
hist(getEnrichment(enrichment_H3K27me3))
```

## Calling Differential Enrichment between two Conditions

```{r}
difference <- diffR(getCounts(enrichment_H3K27me3)$treatment,
                    getCounts(enrichment_H3K4me3)$treatment,
                    enrichment_H3K27me3@ranges)
```

## Identify Regimes of Enrichment

```{r}
difference <- diffR(getCounts(enrichment_H3K4me3)$control,
                    getCounts(enrichment_H3K4me3)$treatment,
                    enrichment_H3K27me3@ranges)
```

# Advanced Topics

##
